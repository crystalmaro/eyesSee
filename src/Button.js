/* eslint-disable */
import React, { Component } from 'react';
import './css/button.css';

class Button extends Component {
  render() {
    return (
      <div className="buttonContainer">
        <div className="button">COMPARE</div>
        <div className="button">NEXT</div>
      </div>
    );
  }
}

export default Button;
