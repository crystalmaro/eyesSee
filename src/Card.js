// import React, { Component } from 'react'

// class Card extends Component {
//   render() {
//     return (
//       <div
//         onClick={this.props.clickImg}

//         style={this.props.yesOnTop == true ? { zIndex: 1 } : null}
//         className={this.props.origClass}
//         id={this.props.isYes}
//         level={this.props.imgData[this.props.randomRound[this.props.currentRound]].level}
//       >
//         <img src={this.props.imgData[this.props.randomRound[this.props.currentRound]].this.props.isYes} alt="" />
//       </div>
//     )
//   }
// }

// export default Card;

import React from 'react';

const Card = (clickImg, yesOnTop, origClass, imgData, randomRound, currentRound, isYes) => {
  return (
    <div
      onClick={clickImg}

      style={yesOnTop == true ? { zIndex: 1 } : null}
      className={origClass}
      id={isYes}
      level={imgData[randomRound[currentRound]].level}
    >
      <img src={imgData[randomRound[currentRound]].isYes} alt="" />
    </div>
  );
}

export default Card;
