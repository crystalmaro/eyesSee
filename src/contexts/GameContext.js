import React, { createContext, Component } from 'react';
import firebase from 'Firebase';
import '@firebase/firestore';

export const GameContext = createContext();

class GameContextProvider extends Component {
	state = {
		origClass: 'imgBox',
		compareClass: 'button',
		scoreClass: 'score',
		isClicked: false,
		yesOnTop: true,
		isCorrect: true,
		yesImg: './imageStock/yes.png',
		noImg: './imageStock/no.png',
		currentRound: 0,
		currentScore: 0,
		progressBar: 0,
		isResultReady: false,
		ranking: 0,
		rankingMes: '',
		// ===== Firebase imageStock data
		imgData: [], // all data loaded from firebase
		totalRound: [], // array of number 0 - 19
		randomRound: [], // randomized EasyRound, 1 followed by randomize HardRound
		// ===== Firebase masterScore data
		globalScoreArray: [],
		timer: 0,
		isTimerOn: true
	};

	// ============================
	// load firebase data - masterScore & imageStock
	// ============================
	loadFirebase = (e) => {
		const db = firebase.firestore();
		// ===== masterScore
		db.collection('masterScore').doc('scoreInfo').get().then((doc) => {
			if (doc.exists) {
				// let obj = doc.data();
				// for (const key in obj) {
				// 	console.log(`${key} -> ${obj[key]}`);
				// }

				this.setState({
					// ===== NEW Firebase masterScore data (ARRAY)
					globalScoreArray: doc.data().scoreData.sort(function(a, b) {
						return b - a;
					})
				});
			}
		});

		// ===== imageStock
		db.collection('imageStock').get().then((querySnapshot) => {
			querySnapshot.forEach((doc) => {
				// 1. load and save firebase data into state
				let imgData = [ ...this.state.imgData, doc.data() ];
				// 2. generate a number array for the length of this.state.imgData
				let totalRound = [ ...this.state.totalRound, this.state.imgData.length ];
				this.setState({ imgData, totalRound });
			});
			// 3. slice() returns a new array, without mutating the original
			let totalEasyRound = this.state.totalRound.slice(0, 10);
			let totalHardRound = this.state.totalRound.slice(10);
			// 4.0 generate randomized index order for array

			if (this.state.randomRound.length < 20) {
				// 4.1 - EASY
				for (let a = totalEasyRound, i = a.length; i--; ) {
					let randomRound = [
						...this.state.randomRound,
						a.splice(Math.floor(Math.random() * (i + 1)), 1)[0]
					];
					this.setState({ randomRound });
				}
				// 4.2 - HARD
				for (let a = totalHardRound, i = a.length; i--; ) {
					let randomRound = [
						...this.state.randomRound,
						a.splice(Math.floor(Math.random() * (i + 1)), 1)[0]
					];
					this.setState({ randomRound });
				}
			}
			console.log(this.state.randomRound);
		});
	};

	// ============================
	// UI events - click image
	// ============================
	clickImg = (e) => {
		this.setState({ isTimerOn: false });
		// 1. overlap both images & show current round answer
		this.overlapImages(e);
		// 2. decide which image on top & decide current round answer
		this.checkImage(e);
		// 3. scorekeeping
		this.addScore(e);
		// 4. progress bar incrementation
		this.addProgressBar(e);
		// 5. NEXT button changes to RESULT on last round, after image selection
		if (this.state.currentRound == this.state.randomRound.length - 1) {
			this.setState({ isResultReady: true });
			// if (this.state.isResultReady) {
			this.calculateRanking(e);
			// }
		}
	};
	overlapImages = (e) => {
		this.setState({
			origClass: 'imgBox clicked',
			isClicked: true,
			compareClass: 'button isClicked'
		});
	};
	checkImage = (e) => {
		switch (e.target.parentNode.id) {
			case 'no':
				this.setState({
					yesOnTop: false,
					isCorrect: false
				});
				break;
			case 'yes':
				this.setState({
					yesOnTop: true,
					isCorrect: true,
					scoreClass: 'score yes'
				});
				break;
		}
	};
	addScore = (e) => {
		if (e.target.parentNode.id == 'yes' && !this.state.isClicked) {
			switch (e.target.parentNode.getAttribute('level')) {
				case 'Easy':
					this.setState((preState) => {
						return { currentScore: preState.currentScore + 100 };
					});
					break;
				case 'Hard':
					this.setState((preState) => {
						return { currentScore: preState.currentScore + 150 };
					});
					break;
			}
		}
	};
	addProgressBar = (e) => {
		// TODO change progressBar width from px to %
		if (!this.state.isClicked && this.state.progressBar < 100) {
			this.setState((preState) => {
				return { progressBar: preState.progressBar + 100 / 20 };
			});
		}
	};
	calculateRanking = (e) => {
		// 1. calculate accuracy rate
		let point = Math.round(this.state.currentScore / 3000 * 100);
		// 2. add current score into array
		let newArr = this.state.globalScoreArray.slice();
		newArr.push(point);
		// 3. sort the array in descending order
		newArr.sort(function(a, b) {
			return b - a;
		});
		// 4. final score = arr.indexOf(percentage) / arr.length
		let index = Number(newArr.indexOf(point)) + 1;
		let ranking = Math.round(index / newArr.length * 100);
		console.log('rank:' + ranking);
		switch (true) {
			case 0 < ranking < 10:
				var rankingMsg = "Here's the Golden Ticket to join my developing team!";
				break;
			case 10 < ranking < 20:
				var rankingMsg = "My grandma's eyes are better than yours, just saying.";
				break;
			case 20 < ranking < 40:
				var rankingMsg = 'You need to get your eyes checked, like seriously.';
				break;
			case 40 < ranking < 60:
				var rankingMsg = 'Might as well donate your eyes.';
				break;
			case 60 < ranking < 80:
				var rankingMsg = 'Do you think this is a guessing game, hun?!';
				break;
			case 80 < ranking < 100:
				var rankingMsg = "Excellent job!! (No, not really you're horrible)";
				break;
			default:
				var rankingMsg = 'Play again!';
		}
		// 5. save the ranking to state, and pass the ranking to <Result />
		this.setState({
			ranking,
			rankingMsg
		});
		// 6. save the newly updated array back to firebase
		const db = firebase.firestore();
		return db.collection('masterScore').doc('scoreInfo').set({
			scoreData: newArr
		});
	};

	// ============================
	// UI events - click next
	// ============================
	clickNext = (e) => {
		this.resetRound(e);
		this.setState({ isTimerOn: true });
	};
	resetRound = (e) => {
		if (this.state.isClicked) {
			switch (this.state.currentRound == this.state.currentRound.length - 1) {
				case false:
					this.setState(
						(preState) => {
							// increment current round count
							return { currentRound: preState.currentRound + 1 };
						},
						() => {
							this.setState({
								// reset state for the next round
								origClass: 'imgBox',
								isClicked: false,
								yesOnTop: true,
								isCorrect: true,
								compareClass: 'button',
								scoreClass: 'score'
							});
							// console.log('inside : ' + this.state.currentRound);
						}
					);
					// console.log('outside : ' + this.state.currentRound);

					break;
				// case true:
				// 	break;
			}
			return;
		}
	};

	// ============================
	// UI events - click compare
	// ============================
	compareMouseDown = (e) => {
		if (this.state.isClicked) {
			this.setState({ compareClass: 'button compare' });
			switch (this.state.yesOnTop) {
				case true:
					this.setState({ yesOnTop: false, isCorrect: false });
					break;
				case false:
					this.setState({ yesOnTop: true, isCorrect: true });
					break;
			}
		}
	};
	compareMouseUp = (e) => {
		if (this.state.isClicked) {
			this.setState({ compareClass: 'button isClicked' });
			switch (this.state.yesOnTop) {
				case true:
					this.setState({ yesOnTop: false, isCorrect: false });
					break;
				case false:
					this.setState({ yesOnTop: true, isCorrect: true });
					break;
			}
		}
	};
	// ============================
	// timer - stopwatch
	// ============================
	// ! below is reserved for Stopwatch (class component)
	// ! but running the timer every second,
	// ! changes the state within GameContext,
	// ! when <Game> detects changes within render/return
	// ! as being wrapped by <GameContext.Consumer>
	// ! it re-renders and re-run the random function within <Card>
	getSeconds = () => {
		return ('0' + this.state.timer % 60).slice(-2);
	};
	getMinutes = () => {
		return Math.floor(this.state.timer / 60);
	};
	setTimer = () => {
		this.setState((preState) => {
			return { timer: preState.timer + 1 };
		});
	};

	render() {
		return (
			<GameContext.Provider
				value={{
					...this.state,
					clickImg: this.clickImg,
					compareMouseDown: this.compareMouseDown,
					compareMouseUp: this.compareMouseUp,
					clickNext: this.clickNext,
					loadFirebase: this.loadFirebase,
					getSeconds: this.getSeconds,
					getMinutes: this.getMinutes,
					setTimer: this.setTimer
				}}
			>
				{this.props.children}
			</GameContext.Provider>
		);
	}
}

export default GameContextProvider;
