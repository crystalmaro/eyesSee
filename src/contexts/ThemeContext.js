import React, { createContext, Component } from 'react';

export const ThemeContext = createContext();

class ThemeContextProvider extends Component {
	state = {
		isLightTheme: true,
		light: { font: '#595959', bg: '#eef2f3', ui: '#ffb7b0', logo: './imageStock/595959.png' },
		dark: { font: '#eef2f3', bg: '#595959', ui: 'green', logo: './imageStock/eef2f3.png' }
	};
	toggleTheme = () => {
		this.setState({ isLightTheme: !this.state.isLightTheme });
	};
	render() {
		return (
			<ThemeContext.Provider
				value={{
					...this.state,
					toggleTheme: this.toggleTheme
				}}
			>
				{this.props.children}
			</ThemeContext.Provider>
		);
	}
}

export default ThemeContextProvider;
