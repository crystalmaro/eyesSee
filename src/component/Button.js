import React, { Component } from 'react';
import { GameContext } from '../contexts/GameContext';
import { Route, Router, BrowserRouter, NavLink } from 'react-router-dom';
import Result from './Result';

class Button extends Component {
	render() {
		return (
			// <BrowserRouter>
			<GameContext.Consumer>
				{(context) => {
					const {
						compareMouseDown,
						compareMouseUp,
						compareClass,
						clickNext,
						isResultReady,
						isClicked
					} = context;
					return (
						<div className="buttonContainer">
							<div
								onMouseDown={compareMouseDown}
								onTouchStart={compareMouseDown}
								onMouseUp={compareMouseUp}
								onTouchEnd={compareMouseUp}
								className={compareClass}
							>
								COMPARE
							</div>
							<div onClick={clickNext} className={isClicked ? 'button isClicked' : 'button'}>
								{isResultReady ? (
									<NavLink to="/result">
										<strong>RESULT</strong>
									</NavLink>
								) : (
									'NEXT'
								)}
							</div>
						</div>
					);
				}}
			</GameContext.Consumer>
			// </BrowserRouter>
		);
	}
}

export default Button;
