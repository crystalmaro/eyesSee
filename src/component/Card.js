import React, { Component } from 'react';
import { GameContext } from '../contexts/GameContext';

class Card extends Component {
	render() {
		return (
			<GameContext.Consumer>
				{(context) => {
					const { clickImg, yesOnTop, origClass, currentRound, imgData, randomRound } = context;
					let first;
					let second;

					const yes = (
						<div
							onClick={clickImg}
							style={yesOnTop ? { zIndex: 1 } : null}
							className={origClass}
							id="yes"
							level={imgData[randomRound[currentRound]].level}
						>
							<img src={`${imgData[randomRound[currentRound]].yes}`} alt="" />
						</div>
					);

					const no = (
						<div
							onClick={clickImg}
							style={yesOnTop ? null : { zIndex: 1 }}
							className={origClass}
							id="no"
							level={imgData[randomRound[currentRound]].level}
						>
							<img src={imgData[randomRound[currentRound]].no} alt="" />
						</div>
					);

					switch (Math.random() > 0.5) {
						case true:
							first = yes;
							second = no;
							break;
						case false:
							first = no;
							second = yes;
							break;
					}

					return (
						<div className="gameContainer" id={randomRound[currentRound]}>
							{first}
							{second}
						</div>
					);
				}}
			</GameContext.Consumer>
		);
	}
}

export default Card;
