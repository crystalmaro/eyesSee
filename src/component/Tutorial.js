import React, { Component } from 'react';
import { Link, Route } from 'react-router-dom';
import Intro from './Intro';
import Game from './Game';
import { ThemeContext } from '../contexts/ThemeContext';
import ThemeToggle from './ThemeToggle';
import { Steps, Hints } from 'intro.js-react';
import 'intro.js/introjs.css';
import '../css/tutorial.css';
import Stopwatch from './Stopwatch';

class Tutorial extends Component {
	state = {
		introData: [
			{
				key: 1,
				yes: './imageStock/intro1.png',
				no: './imageStock/intro2.png',
				reason: 'Click and hold compare to see the difference'
			},
			{
				key: 2,
				yes: './imageStock/intro3.png',
				no: './imageStock/intro4.png',
				reason: 'Are you ready?'
			}
		],
		introRound: 0,
		introMsg: [ 'Which UI is more correct?', 'Please read the rules below.' ],

    origClass: 'imgBox',
		compareClass: 'button',
		isClicked: false,
		yesOnTop: true,
		isCorrect: true,
		yesImg: './imageStock/yes.png',
		noImg: './imageStock/no.png',
		isIntroDone: false
		// ! ===== below isnt currently being used
		// min: 1,
		// sec: 0
	};

	// ============================
	// UI events - click image
	// ============================
	clickImg = (e) => {
		this.overlapImages(e);
		this.checkImageSelection(e);
		// this.setState({ introMsg: 'Please read the rules below.' });
		if (this.state.introRound == this.state.introData.length - 1) {
			this.setState({ isIntroDone: true });
			return;
		}
	};
	overlapImages = (e) => {
		this.setState({
			origClass: 'imgBox clicked',
			isClicked: true,
			compareClass: 'button isClicked'
		});
	};
	checkImageSelection = (e) => {
		switch (e.target.parentNode.id) {
			case 'no':
				this.setState({
					yesOnTop: false,
					isCorrect: false
				});
				break;
			case 'yes':
				this.setState({
					yesOnTop: true,
					isCorrect: true
				});
				break;
		}
	};
	// ============================
	// UI events - click next
	// ============================
	clickNext = (e) => {
		this.resetRound(e);
	};
	resetRound = (e) => {
		if (this.state.isClicked && this.state.introRound !== this.state.introRound.length - 1) {
			this.setState(
				(preState) => {
					return { introRound: preState.introRound + 1 };
				},
				// callback
				() => {
					this.setState({
						origClass: 'imgBox',
						isClicked: false,
						yesOnTop: true,
						isCorrect: true,
						compareClass: 'button'
					});
				}
			);
			return;
		}
	};
	// ============================
	// UI events - click compare
	// ============================
	compareMouseDown = (e) => {
		if (this.state.isClicked) {
			this.setState({ compareClass: 'button compare' });
			switch (this.state.yesOnTop) {
				case true:
					this.setState({ yesOnTop: false, isCorrect: false });
					break;
				case false:
					this.setState({ yesOnTop: true, isCorrect: true });
					break;
			}
			return;
		}
	};
	compareMouseUp = (e) => {
		if (this.state.isClicked) {
			this.setState({ compareClass: 'button isClicked' });
			switch (this.state.yesOnTop) {
				case true:
					this.setState({ yesOnTop: false, isCorrect: false });
					break;
				case false:
					this.setState({ yesOnTop: true, isCorrect: true });
					break;
			}
			return;
		}
	};

	render() {
		return (
			<ThemeContext.Consumer>
				{(context) => {
					const { isLightTheme, light, dark, stepsEnabled, steps, onExit } = context;
					const theme = isLightTheme ? light : dark;
					return (
						<div className="tutorialContainer">
							<Steps enabled={stepsEnabled} steps={steps} initialStep={0} onExit={onExit} />
							<div className="tutorialHeader">
								<ThemeToggle />
								<div
									className="introMsg selectOff"
									style={this.state.isClicked ? { display: 'none' } : { display: 'block' }}
								>
									{this.state.introMsg[this.state.introRound]}
								</div>
							</div>

							<div className="gameContainer">
								<div
									onClick={this.clickImg}
									style={this.state.yesOnTop ? { zIndex: 1 } : null}
									className={this.state.origClass}
									id="yes"
								>
									<img src={this.state.introData[this.state.introRound].yes} alt="" />
								</div>
								<div
									onClick={this.clickImg}
									style={this.state.yesOnTop ? null : { zIndex: 1 }}
									className={this.state.origClass}
									id="no"
								>
									<img src={this.state.introData[this.state.introRound].no} alt="" />
								</div>
							</div>

							<div
								className="messageContainer"
								style={this.state.isClicked ? { display: 'flex' } : { display: 'none' }}
							>
								<div className="result">
									<img src={this.state.isCorrect ? this.state.yesImg : this.state.noImg} />
									<div className="message">{this.state.introData[this.state.introRound].reason}</div>
								</div>
							</div>

							<div
								className="buttonContainer"
								style={this.state.isClicked ? { display: 'flex' } : { display: 'none' }}
							>
								<div
									onMouseDown={this.compareMouseDown}
									onTouchStart={this.compareMouseDown}
									onMouseUp={this.compareMouseUp}
									onTouchEnd={this.compareMouseUp}
									className={this.state.compareClass}
								>
									COMPARE
								</div>
								<div
									onClick={this.clickNext}
									className={this.state.isClicked ? 'button isClicked' : 'button'}
								>
									{this.state.isIntroDone ? (
										<Link to="/game">
											<strong>START GAME</strong>
										</Link>
									) : (
										'NEXT'
									)}
								</div>
							</div>
						</div>
					);
				}}
			</ThemeContext.Consumer>
		);
	}
}

export default Tutorial;
