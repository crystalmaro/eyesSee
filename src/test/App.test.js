import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { GameContext } from '../contexts/GameContext';
import Adapter from 'enzyme-adapter-react-16';
import Tutorial from '../component/Tutorial';
import Card from '../component/Card';
import Button from '../component/Button';

Enzyme.configure({ adapter: new Adapter() });

it('<Button> component renders correctly', () => {
	const tree = renderer
		.create(
			<GameContext.Provider
				value={{
					isResultReady: false,
					compareClass: 'button'
				}}
			>
				<Button />
			</GameContext.Provider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

describe('enzyme context snapshot for <Card>', () => {
	it('should render children for snapshot', () => {
		const tree = shallow(<Card />);
		expect(tree).toMatchSnapshot();
	});
});

describe('Tutorial component', () => {
	const wrapper = shallow(<Tutorial />);
	it('introRound starts at 0', () => {
		const countState = wrapper.state().introRound;
		expect(countState).toEqual(0);
	});
	it('introData[0].key to be 1', () => {
		const dataKey = wrapper.state().introData[0].key;
		expect(dataKey).toBe(1);
	});
	it('introData yes img src(object key value pair)', () => {
		const introDataYesImg = wrapper.state().introData[0].yes;
		expect(introDataYesImg).toEqual(expect.stringContaining('./imageStock/intro1.png'));
	});
	it('introData yes img src(string)', () => {
		const introDataYesImg = wrapper.state().introData[0];
		expect(introDataYesImg).toEqual(expect.objectContaining({ yes: './imageStock/intro1.png' }));
	});
	it('introMsg length of 2', () => {
		const introMsg = wrapper.state().introMsg;
		expect(introMsg).toHaveLength(2);
	});
	it('isIntroDone to be falsy', () => {
		const isIntroDone = wrapper.state().isIntroDone;
		expect(isIntroDone).toBeFalsy();
	});
});
