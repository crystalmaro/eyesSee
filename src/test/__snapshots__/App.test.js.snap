// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> component renders correctly 1`] = `
Array [
  <div
    className="helloResult"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Great job! Result calculating...
  </div>,
  <div
    className="buttonContainer"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="button"
    >
      COMPARE
    </div>
    <div
      className="button"
    >
      NEXT
    </div>
  </div>,
]
`;

exports[`enzyme context snapshot for <Card> should render children for snapshot 1`] = `ShallowWrapper {}`;
